src/merkle.cpp

```diff
@@ -208,5 +208,34 @@ namespace libtorrent {
 		return ret;
 	}
 
+	std::pair<aux::vector<std::pair<sha256_hash, sha256_hash>>, sha256_hash>
+	merkle_check_proofs(int proof_leafs, sha256_hash tree_root, span<sha256_hash const> hashes
+		, int const index)
+	{
+		aux::vector<std::pair<sha256_hash, sha256_hash>> ret(int(hashes.size()));
+		auto ret_it = ret.begin();
+		for (auto const& proof : hashes)
+		{
+			proof_leafs *= 2;
```

Comment:
@ssiloti I'm working on unit tests for this function, and this line seems wrong.

my understanding of this snippet of code is that `tree_root` is the root of the sub tree we just inserted. We've validated that the inserted sub tree is internally consistent, now we want to make sure its root is valid in the full tree we're inserting it into.

In this context, `tree_root` is really at the bottom of part of the tree we're interested in. We'll iterate *up* the tree while using the uncle hashes passed in via `hashes`. At the end we have the root of the full tree (or at least the hash of a known-good node in the full tree) which is returned, along with the hashes of the nodes along the path.

The caller can then compare the returned hash against the known good one and determine whether `tree_node` was correct.

Given this understanding, it seems this line `proof_leafs *= 2` is wrong, and should be `proof_leafs /= 2`. Since we traverse *up* the tree, each layer has fewer nodes. In fact, I would expect a simpler way to implement this would be to skip the `proof_leafs` entirely, and just halve the `index` here (`index /= 2`) and `proof_right = (index & 1) == 0;`

Those changes are consistent with my understanding and it makes my test pass, but I'm worried I'm missing something.

Reply:
it's possible I had misunderstood the meaning of `index`. It seems to be the offset from the start of the layer of the first inserted *leaf*, not the layer offset for the *root*. This code is only concerned about the root though, so I think it would be a simplification to make it refer to the offset of `tree_root`.

Reply:
D'oh, you've got it right. Not sure how this code worked at all in my testing. I agree it would be simpler to use the index of the root to determine the positions of the sibling hashes.

Reply:
I broke this out into a separate PR: https://github.com/arvidn/libtorrent/pull/4652